docs/homepage/blog/ospp_mid-term_report_210370190/index.md

```diff
@@ -0,0 +1,400 @@
+        rng = rng,
+        :state => Vector{Int},
+        :action => Int,
+    ),
+)
+```
+
+Based on the defined agents, the `NFSPAgentManager` can be customized as the following:
+```Julia
+nfsp = NFSPAgentManager(
+    Dict(
+        (player, NFSPAgent(
+            deepcopy(rl_agent),
+            deepcopy(sl_agent),
+            0.1f0, # anticipatory parameter
+            rng,
+            128, # update_freq
+            0, # initial update_step
+            true, # initial NFSPAgent's learn mode
+        )) for player in players(wrapped_env) if player != chance_player(wrapped_env)
+    )
+)
+```
+
+Based on the setting [`stop_condition`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/docs/experiments/experiments/NFSP/JuliaRL_NFSP_KuhnPoker.jl#L126) and designed [`hook`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/docs/experiments/experiments/NFSP/JuliaRL_NFSP_KuhnPoker.jl#L15), you can just `run(nfsp, wrapped_env, stop_condition, hook)` to perform the experiment. Use [`Plots.plot`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/66a9c7e41416c4b959c66b07d06d67915dc4b628/docs/experiments/experiments/NFSP/JuliaRL_NFSP_KuhnPoker.jl#L136) to get the following result: (here [`nash_conv`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/639717388fb41199c98b90406bea76232bc6294d/src/ReinforcementLearningZoo/src/algorithms/cfr/nash_conv.jl#L1) is one common metric to show the performance of a multi-agent reinforcement learning algorithm.)
+
+\dfig{body;JuliaRL_NFSP_KuhnPoker.png;Result of the experiment.}
+
+### 2.3 Multi-agent Deep Deterministic Policy Gradient(MADDPG) algorithm
+
+#### Brief Introduction
+
+The Multi-agent Deep Deterministic Policy Gradient(MADDPG)\dcite{DBLP:journals/corr/LoweWTHAM17} algorithm improves the [Deep Deterministic Policy Gradient(DDPG)](https://spinningup.openai.com/en/latest/algorithms/ddpg.html), which works well on multi-agent games. Based on the DDPG, the critic of each agent in MADDPG can get all agents' policies according to the paper's hypothesis\dcite{DBLP:journals/corr/LoweWTHAM17}, including their personal states and actions, which can help get a more reasonable score of the actor's policy.
+
+#### Implementation
+
+Since there has been [`DDPGPolicy`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/ddpg.jl#L47) in the RLZoo, I implement the [`MADDPGManager`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl) which is a special multi-agent manager that all agents apply `DDPGPolicy` with one **improved critic**. The structure of `MADDPGManager` is as the following:
```

Comment:
```suggestion
Given that the [`DDPGPolicy`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/50756bbe9e1925a9320d1abdbbc6255c1b4a27f1/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/ddpg.jl#L47) is already implemented in the RLZoo, I implement the [`MADDPGManager`](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/master/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl) which is a special multi-agent manager that all agents apply `DDPGPolicy` with one **improved critic**. The structure of `MADDPGManager` is as the following:
```

Reply:
Thanks for your patience in reviewing and sorry for my terrible report. 😞 I'll update my report follow your advice soon!